src/sinks/datadog/logs.rs

```diff
@@ -66,31 +130,223 @@ impl SinkConfig for DatadogLogsConfig {
+impl HttpSink for DatadogLogsTextService {
+    type Input = Bytes;
+    type Output = Vec<Bytes>;
 
-    if let Some(timestamp) = log.remove(&log_schema().timestamp_key()) {
-        log.insert("date", timestamp);
+    fn encode_event(&self, event: Event) -> Option<Self::Input> {
+        encode_event(event, &self.config.encoding)
     }
 
-    if let Some(host) = log.remove(&log_schema().host_key()) {
-        log.insert("host", host);
+    async fn build_request(&self, events: Self::Output) -> crate::Result<http::Request<Vec<u8>>> {
+        let body: Vec<u8> = events.iter().flat_map(|b| b.into_iter()).cloned().collect();
+
+        build_request(&self.config, body)
     }
+}
+
+/// Build the request, GZipping the contents if the config specifies.
+fn build_request(
+    config: &DatadogLogsConfig,
+    body: Vec<u8>,
+) -> crate::Result<http::Request<Vec<u8>>> {
+    let uri = config.get_endpoint();
+    let request = Request::post(uri)
+        .header("Content-Type", "text/plain")
+        .header("DD-API-KEY", config.api_key.clone());
+
+    let (request, body) = match config.compression {
+        Compression::None => (request, body),
+        Compression::Gzip => {
+            let mut encoder = GzEncoder::new(
+                Vec::new(),
+                match config.compression_level {
+                    Some(level) if level <= 9 => flate2::Compression::new(level),
+                    _ => flate2::Compression::fast(),
+                },
+            );
+
+            encoder.write_all(&body)?;
+            (
+                request.header("Content-Encoding", "gzip"),
+                encoder.finish()?,
+            )
+        }
+    };
+
+    request
+        .header("Content-Length", body.len())
+        .body(body)
+        .map_err(Into::into)
+}
+
+/// The healthcheck is performed by sending an empty request to Datadog and checking
+/// the return.
+async fn healthcheck<T, O>(config: T, mut client: HttpClient) -> crate::Result<()>
+where
+    T: HttpSink<Output = Vec<O>>,
+{
+    let req = config.build_request(Vec::new()).await?.map(Body::from);
+
+    let res = client.send(req).await?;
 
-    if let Some(bytes) = util::encode_event(event, encoding) {
-        // Prepend the api_key:
-        // {API_KEY} {EVENT_BYTES}
-        let api_key = format!("{} {}", api_key, String::from_utf8_lossy(&bytes));
-        Some(Bytes::from(api_key))
+    let status = res.status();
+    let body = hyper::body::to_bytes(res.into_body()).await?;
+
+    if status == StatusCode::OK {
+        Ok(())
+    } else if status == StatusCode::UNAUTHORIZED {
+        let json: serde_json::Value = serde_json::from_slice(&body[..])?;
+
+        let message = if let Some(s) = json
+            .as_object()
+            .and_then(|o| o.get("error"))
+            .and_then(|s| s.as_str())
+        {
+            s.to_string()
+        } else {
+            "Token is not valid, 401 returned.".to_string()
+        };
+
+        Err(message.into())
```

Comment:
```suggestion
        Err(json
            .as_object()
            .and_then(|o| o.get("error"))
            .and_then(|s| s.as_str())
            .unwrap_or("Token is not valid, 401 returned.")
            .to_string()
            .into()
        )
```
Might have to flip the order of the last two lines if there's a lifetime conflict on the `&str`